src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+      mongoc_server_description_handle_ismaster (description, NULL, -1, &error);
+      _server_monitor_heartbeat_failed (
+         server_monitor, &description->error, duration_us, awaited);
+   }
+
+   if (command_or_network_error) {
+      if (server_monitor->stream) {
+         mongoc_stream_failed (server_monitor->stream);
+      }
+      server_monitor->stream = NULL;
+      server_monitor->more_to_come = false;
+      bson_mutex_lock (&server_monitor->topology->mutex);
+      _mongoc_topology_clear_connection_pool (server_monitor->topology,
+                                              server_monitor->description->id);
+      bson_mutex_unlock (&server_monitor->topology->mutex);
+   }
+
+   bson_destroy (&ismaster_reply);
+   return description;
+}
+
+/* Request scan of a single server.
+ *
+ * Caller does not need to have topology mutex locked.
+ * Locks server monitor mutex to deliver scan_requested.
+ */
+void
+mongoc_server_monitor_request_scan (mongoc_server_monitor_t *server_monitor)
+{
+   MONITOR_LOG (server_monitor, "requesting scan");
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.scan_requested = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* Request cancellation of an in progress awaitable ismaster.
+ *
+ * Called from app threads on network errors and during shutdown.
+ * Caller must not hold lock.
+ * Locks server monitor mutex.
+ */
+void
+mongoc_server_monitor_request_cancel (mongoc_server_monitor_t *server_monitor)
+{
+   MONITOR_LOG (server_monitor, "requesting cancel");
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.cancel_requested = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* Wait for heartbeatFrequencyMS or minHeartbeatFrequencyMS if a scan is
+ * requested.
+ *
+ * Locks server monitor mutex.
+ */
+void
+mongoc_server_monitor_wait (mongoc_server_monitor_t *server_monitor)
+{
+   int64_t start_ms;
+   int64_t now_ms;
+   int64_t scan_due_ms;
+
+   start_ms = _now_ms ();
+   scan_due_ms = start_ms + server_monitor->heartbeat_frequency_ms;
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   while (true) {
+      int64_t sleep_duration_ms;
+
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         break;
+      }
+
+      now_ms = _now_ms ();
+      if (now_ms >= scan_due_ms) {
+         break;
+      }
+
+      if (server_monitor->shared.scan_requested) {
+         server_monitor->shared.scan_requested = false;
+         scan_due_ms = start_ms + server_monitor->min_heartbeat_frequency_ms;
+      }
+
+
+      sleep_duration_ms = scan_due_ms - now_ms;
+
+      if (sleep_duration_ms > 0) {
+         MONITOR_LOG (
+            server_monitor, "sleeping for %" PRId64, sleep_duration_ms);
+         mongoc_cond_timedwait (&server_monitor->shared.cond,
```

Comment:
Should we be concerned about a non-successful return? Is there a case where this returns and the mutex is _not_ re-locked? See [`pthread_cond_timedwait`](https://linux.die.net/man/3/pthread_cond_timedwait) and the alternative definition in `mongoc-thread-private.h` for Windows.

If this results in adding more logic here, I might suggest refactoring the `if/else` to just have an early guard statement that breaks and allows us to leave this code at the top level of the `while` iteration. For example:

```c
if (sleep_duration_ms <= 0) {
   break;
}

MONITOR_LOG...
```

Reply:
Handling timeout errors works for me. Thanks.